File: python/src/accuracy_reporter.py
from csv_chart_drawer import CSVChartDrawer
from accuracy_check_query import AccuracyCheckQuery
import os
import datetime
import sys
sys.path.append('./src/lib')
sys.path.append('./src/queries')


class AccuracyReporter:
    def __init__(
            self,
            scheme: str,
            host: str,
            bot_id: str,
            user_id: str,
            test_data: str) -> None:
        self.scheme = scheme
        self.host = host
        self.bot_id = bot_id
        self.user_id = user_id
        self.test_data = test_data
        self.accuracy_check_query = AccuracyCheckQuery(
            self.scheme, self.host, self.bot_id, self.user_id, self.test_data)
        self.res_bodies = self.accuracy_check_query.res_bodies()
        self.csv_chart_drawer = CSVChartDrawer(self.test_data, self.res_bodies)

    def run(self, dirname: str) -> None:
        with open(self.__filename__(dirname), 'w') as f:
            self.csv_chart_drawer.csv(f)

    # private

    def __filename__(self, dirname: str) -> str:
        return os.path.join(
            dirname, f'accuracy_score_chart_{
                datetime.datetime.now():%Y%m%d%H%M%S}.csv'
        )

File: ruby/src/accuracy_reporter.rb
# rbs_inline: enabled

require_relative './queries/accuracy_check_query'
require_relative './lib/csv_chart_drawer'

class AccuracyReporter
  # @rbs scheme: String
  # @rbs host: String
  # @rbs bot_id: String
  # @rbs user_id: String
  # @rbs path_to_test_data: String
  # @rbs path_to_accuracy_score_chart: String
  # @rbs return: void
  def self.run(scheme:, host:, bot_id:, user_id:, path_to_test_data:, path_to_accuracy_score_chart:)
    new(scheme:, host:, bot_id:, user_id:, path_to_test_data:, path_to_accuracy_score_chart:).run
  end

  # @rbs scheme: String
  # @rbs host: String
  # @rbs bot_id: String
  # @rbs user_id: String
  # @rbs path_to_test_data: String
  # @rbs path_to_accuracy_score_chart: String
  # @rbs return: void
  def initialize(scheme:, host:, bot_id:, user_id:, path_to_test_data:, path_to_accuracy_score_chart:)
    @scheme                       = scheme
    @host                         = host
    @bot_id                       = bot_id
    @user_id                      = user_id
    @path_to_test_data            = path_to_test_data
    @path_to_accuracy_score_chart = path_to_accuracy_score_chart
  end

  # @rbs return: void
  def run
    res_bodies = accuracy_check_query.res_bodies
    csv_chart  = csv_chart_drawer(res_bodies).csv
    File.open(filename(path_to_accuracy_score_chart), 'w') { |f| f.puts(csv_chart) }
  end

  private

  attr_reader :scheme, :host, :bot_id, :user_id, :path_to_test_data, :path_to_accuracy_score_chart

  # @rbs return: Queries::AccuracyCheckQuery
  def accuracy_check_query
    @accuracy_check_query ||= ::Queries::AccuracyCheckQuery.new(scheme, host, bot_id, user_id, path_to_test_data)
  end

  # @rbs res_bodies: Array[Hash[String, untyped]]
  # @rbs return: Lib::CsvChartDrawer
  def csv_chart_drawer(res_bodies)
    @csv_chart_drawer ||= ::Lib::CsvChartDrawer.new(path_to_test_data, res_bodies)
  end

  # @rbs path_to_accuracy_score_chart: String
  # @rbs return: String
  def filename(path_to_accuracy_score_chart)
    File.join(path_to_accuracy_score_chart, "accuracy_score_chart_#{Time.now.strftime('%F%T').gsub(/[:\-]/, '')}.csv")
  end
end
